tools/state-viewer/src/commands.rs

```diff
@@ -896,6 +898,207 @@ pub(crate) fn print_epoch_info(
+        BTreeMap::from_iter(epoch_infos.values().map(|e| (e.epoch_height(), e.clone())));
+    let min_epoch_height = epoch_height;
+    let max_epoch_height = *epoch_heights_to_infos.keys().max().unwrap();
+    let epoch_heights_to_validator_infos =
+        BTreeMap::from_iter(epoch_heights_to_ids.iter().filter_map(|(epoch_height, epoch_id)| {
+            if *epoch_height >= min_epoch_height
+                && *epoch_height <= max_epoch_height.saturating_sub(4)
+            {
+                Some((epoch_height, epoch_manager.get_epoch_validator_info(&epoch_id).unwrap()))
+            } else {
+                None
+            }
+        }));
+
+    // Next epoch info and next next epoch config are going to be used for
+    // next next epoch info generation. For convenience of the code to make
+    // two modes work, we modify them on each iteration.
+    let mut next_epoch_info =
+        epoch_heights_to_infos.get(&min_epoch_height.saturating_add(1)).unwrap().as_ref().clone();
+    let mut next_next_epoch_config =
+        epoch_manager.get_config_for_protocol_version(PROTOCOL_VERSION).unwrap();
+
+    // Print data header.
+    match mode {
+        EpochAnalysisMode::CheckConsistency => {
+            println!("HEIGHT | VERSION | STATE SYNCS");
+        }
+        EpochAnalysisMode::Backtest => {
+            println!("epoch_height,original_protocol_version,state_syncs,min_validator_num,diff_validator_num,min_stake,diff_stake,rel_diff_stake");
+            // Start from empty assignment for correct number of shards.
+            *next_epoch_info.chunk_producers_settlement_mut() =
+                vec![vec![]; next_next_epoch_config.shard_layout.shard_ids().collect_vec().len()];
+            // This in fact sets maximal number of validators to 100.
+            // Needed because otherwise generation fails at epoch 1327 with
+            // assertion `stake < threshold` in
+            // chain/epoch-manager/src/validator_selection.rs:227:13.
+            // Probably has something to do with extreme case where all
+            // proposals are selected.
+            next_next_epoch_config.validator_selection_config.num_chunk_validator_seats = 100;
+        }
+    }
+
+    // Each iteration will generate and print *next next* epoch info based on
+    // *next* epoch info for `epoch_height`. This follows epoch generation
+    // logic in the protocol.
+    for (epoch_height, epoch_info) in
+        epoch_heights_to_infos.range(min_epoch_height..=max_epoch_height - 4)
+    {
+        let next_epoch_height = epoch_height.saturating_add(1);
+        let next_next_epoch_height = epoch_height.saturating_add(2);
+        let next_epoch_id = epoch_heights_to_ids.get(&next_epoch_height).unwrap();
+        let next_next_epoch_id = epoch_heights_to_ids.get(&next_next_epoch_height).unwrap();
+        let epoch_summary = epoch_heights_to_validator_infos.get(epoch_height).unwrap();
+
+        next_epoch_info = match mode {
```

Comment:
That's a fun way to implement where you have multiple of those checks :) I'm totally fine with it but I'm curious how did you arrive at this and how does it compare to having two methods, one for each mode?

Reply:
That was just bad code, how about now?
These two modes really share a lot of code. The difference is - for the first one, I always take data from DB. For the second, I override some data with previously generated info to simulate new algorithm behaviour. But the rest stays the same, so I don't want to separate impls completely.